Kotlin/src/test/kotlin/com/gildedrose/GildedRoseTest.kt

```diff
@@ -10,10 +10,123 @@ internal class GildedRoseTest {
         val items = listOf(Item("foo", 0, 0))
         val app = GildedRose(items)
         app.updateQuality()
-        assertEquals("fixme", app.items[0].name)
+        assertEquals("foo", app.items[0].name)
+    }
+
+    @Test
+    fun updateQualityDecrementsSellInForNonLegendaryItems() {
+        val items = listOf(
+            Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
+            Item("Backstage passes to a TAFKAL80ETC concert", -15, 20),
+        )
+        val app = GildedRose(items)
+        app.updateQuality()
+        assertEquals(14, app.items[0].sellIn)
+        assertEquals(-16, app.items[1].sellIn)
+    }
+
+    @Test
+    fun updateQualityDoesNotChangeSellInForLegendaryItems() {
+        val items = listOf(
+            Item("Sulfuras, Hand of Ragnaros", -1, 80) //legendary items does not change
+        )
+        val app = GildedRose(items)
+        app.updateQuality()
+        assertEquals(-1, app.items[0].sellIn)
+    }
+
+    @Test
+    fun qualityChangesCorrectlyForNormalItems() {
+        val items = listOf(
+            Item("Elixir of the Mongoose", 5, 7),
+            Item("Dexterity Vest", 6, 0),
+            Item("Dexterity Vest", 0, 10),
+        )
+        val app = GildedRose(items)
+        app.updateQuality()
+        //quality should be decremented by one
+        assertEquals(6, app.items[0].quality)
+        //quality decreases twice as fast for items past sellIn
+        assertEquals(8, app.items[2].quality)
+    }
+
+    @Test
+    fun qualityChangesCorrectlyForLegendaryItems() {
+        val items = listOf(
+            Item("Sulfuras, Hand of Ragnaros", -1, 80),
+        )
+        val app = GildedRose(items)
+        app.updateQuality()
+        //quality should remain the same
+        assertEquals(80, app.items[0].quality)
+    }
 
+    @Test
+    fun qualityChangesCorrectlyForAgedBrie() {
+        val items = listOf(
+            Item("Aged Brie", 2, 10),
+            Item("Aged Brie", 9, 15),
+            Item("Aged Brie", 5, 10),
+        )
+        val app = GildedRose(items)
+        app.updateQuality()
+        //quality should increase
+        assertEquals(11, app.items[0].quality)
+        assertEquals(16, app.items[1].quality)
+        assertEquals(11, app.items[2].quality)
     }
 
+    @Test
+    fun qualityChangesCorrectlyForBackStagePasses() {
```

Comment:
Consider breaking this up to several smaller tests. If something happens that only modifies one bit of this logic, the whole test will break which won't help find the problem.
How granular cane you make it?